flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java

```diff
@@ -153,25 +157,55 @@ public RecordsWithSplitIds<ConsumerRecord<byte[], byte[]>> fetch() throws IOExce
         return recordsBySplits;
     }
 
-    private void markEmptySplitsAsFinished(KafkaPartitionSplitRecords recordsBySplits) {
+    private void markSplitsAsFinished(KafkaPartitionSplitRecords recordsBySplits) {
+        // Some splits are discovered as empty when handling split additions. These splits should be
+        // added to finished splits to clean up states in split fetcher and source reader.
+        markSplitsAsFinished(emptySplits, recordsBySplits);
+
+        // Some splits are removed when handling split changes. These splits should be
+        // added to finished splits to clean up states in split fetcher and source reader.
+        markSplitsAsFinished(removedSplits, recordsBySplits);
+    }
+
+    private void markSplitsAsFinished(
+            Set<String> splits, KafkaPartitionSplitRecords recordsBySplits) {
         // Some splits are discovered as empty when handling split additions. These splits should be
         // added to finished splits to clean up states in split fetcher and source reader.
-        if (!emptySplits.isEmpty()) {
-            recordsBySplits.finishedSplits.addAll(emptySplits);
-            emptySplits.clear();
+        if (!splits.isEmpty()) {
+            recordsBySplits.finishedSplits.addAll(splits);
+            splits.clear();
         }
     }
 
     @Override
     public void handleSplitsChanges(SplitsChange<KafkaPartitionSplit> splitsChange) {
-        // Get all the partition assignments and stopping offsets.
-        if (!(splitsChange instanceof SplitsAddition)) {
+        if (splitsChange instanceof SplitsAddition) {
+            // Get all the partition assignments and stopping offsets.
+            handleSplitsAddition(splitsChange);
+        } else if (splitsChange instanceof SplitsRemoval) {
+            handleSplitsRemoval(splitsChange);
+        } else {
             throw new UnsupportedOperationException(
                     String.format(
                             "The SplitChange type of %s is not supported.",
                             splitsChange.getClass()));
         }
+    }
+
+    private void handleSplitsRemoval(SplitsChange<KafkaPartitionSplit> splitsRemoval) {
```

Comment:
Would it be better to use the concrete SplitsRemoval type here to avoid accidentally passing SplitsAddition?

Reply:
Judgments have been added to ensure that it is an instance of SplitsRemoval. So I think it is ok.